tests/integration-tests.test.ts

```diff
@@ -409,4 +409,142 @@ describe('integration tests', () => {
+      });
+
+      it('continues executing routes in the current chain before redirecting to the next match', () => {
+         const router = new Router();
+
+         let h1, h2Redirects, h3, h4ShouldBeSkipped, g1ShouldBeSkipped, g2Ends;
+
+         h1 = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+
+         h2Redirects = spy((req: Request, _resp: Response, next: NextCallback): void => {
+            req.url = '/goodbye';
+            next();
+         });
+
+         h3 = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+         h4ShouldBeSkipped = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+         g1ShouldBeSkipped = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+         g2Ends = spy((_req: Request, resp: Response): void => { resp.send('goodbye'); });
+
+         router.get('/goodbye', g1ShouldBeSkipped) // skipped because request starts at /hello
+            .get('/hello', h1, h2Redirects, h3) // even though h2 redirects to /goodbye, the chain should still run h3
+            .get('/hello', h4ShouldBeSkipped) // by now the request is for /goodbye, so skip this
+            .get('/goodbye', g2Ends); // this /goodbye handler gets called because of the redirect above
+
+         app.addSubRouter('/path', router);
+
+         testOutcome('GET', '/path/hello', 'goodbye');
+
+         // This should be skipped because it was registered *before* the `h2Redirects`
+         // handler that causes the re-route to '/goodbye'.
+         assert.notCalled(g1ShouldBeSkipped);
+         assert.calledOnce(h1);
+         // This route handler re-routes the request to '/goodbye'.
+         assert.calledOnce(h2Redirects);
+         // This is the last route handler in the `.get('/hello', h1, h2Redirects, h3)`
+         // chain. The main point of this test is to ensure that it's NOT skipped.
+         assert.calledOnce(h3);
+         assert.notCalled(h4ShouldBeSkipped);
+         assert.calledOnce(g2Ends);
+         assert.callOrder(h1, h2Redirects, h3, g2Ends);
+      });
+
+      it('ignores query string params when `request.url` changes to a URL with query params', () => {
+         const handlers = testRoutesWithRedirect((req: Request) => { req.url = '/goodbye?to=you'; });
+
+         expect(handlers.getHelloReq().url).to.strictlyEqual('/hello');
+         // Expect that the query parameters were stripped from the URL
+         expect(handlers.getGoodbyeReq().url).to.strictlyEqual('/goodbye');
+
+         // Expect that the query parameter `to` was ignored
+         expect(handlers.getGoodbyeReq().query.to).to.strictlyEqual(undefined);
+      });
+
+      it('updates path params when `request.url` changes to a URL with different path params', () => {
+         const router1 = new Router(),
+               router2 = new Router(),
+               USER_ID = '1337',
```

Comment:
Oh, don't you feel l33t now? 🤣 

![](https://motherboard-images.vice.com/content-images/contentimage/no-id/1441922605699573.gif)